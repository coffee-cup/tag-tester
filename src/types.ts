export interface MetaTag {
  name?: string;
  property?: string;
  content?: string;
  value?: string;
}

export interface TagResult {
  url: string;
  tags: MetaTag[];
}

export enum FilterType {
  Html,
  OpenGraph,
  Twitter,
}

export interface Settings {
  onlyShowRecommended: boolean;
  syncSimilarTags: boolean;
  filters: FilterType[];
}
