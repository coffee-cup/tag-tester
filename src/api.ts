import { TagResult } from "./types";

export const fetchTags = async (url: string): Promise<TagResult> => {
  const query = `page=${encodeURIComponent(url)}`;

  const res = await fetch(`/api/html?${query}`);
  if (res.status !== 200) {
    const e = await res.json();
    throw new Error(e.message);
  }

  const json = await res.json();
  return json as TagResult;
};
