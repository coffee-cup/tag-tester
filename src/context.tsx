import * as React from "react";
import { validUrl } from "./utils";
import { Results } from "./types";

export interface OGState {
  url: string;
  isUrlError: boolean;
  error: string;
  setUrl: (value: string) => void;
  fetchTags: () => void;
  results: Results | null;
}

const OGContext = React.createContext<OGState>({} as OGState);

export const useOG = (): OGState => {
  const state = React.useContext(OGContext);
  return state;
};

export const OGProvider: React.FC = props => {
  const [url, setUrl] = React.useState("https://alfie.prodo.ai");
  const [error, setError] = React.useState<string | null>(null);
  const [isUrlError, setIsUrlError] = React.useState(false);

  const [results, setResults] = React.useState<Results | null>(null);

  const fetchTags = async () => {
    if (isUrlError || url === "") {
      return;
    }

    const query = `page=${encodeURIComponent(url)}`;
    const json = await fetch(`/api/html?${query}`).then(res => res.json());

    if (json.error != null) {
      setError(json.error);
    } else {
      setResults(json);
      setError(null);
    }
  };

  React.useEffect(() => {
    if (results == null) {
      fetchTags();
    }
  }, []);

  const value: OGState = {
    url,
    isUrlError,
    error,
    setUrl: value => {
      setUrl(value);
      setIsUrlError(value !== "" && !validUrl(value));
    },
    fetchTags,
    results,
  };

  return (
    <OGContext.Provider value={value}>{props.children}</OGContext.Provider>
  );
};
