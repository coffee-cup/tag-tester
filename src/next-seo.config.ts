import { DefaultSeoProps } from "next-seo";

const title = "Tag Tester";
const url = "https://tagtester.dev";
const description =
  "Test your websites html, open graph and twitter metadata tags.";
const image = "https://tagtester.dev/tagtester.png";

const config: DefaultSeoProps = {
  title,
  description,
  openGraph: {
    type: "website",
    url,
    site_name: title,
    images: [
      {
        url: image,
        width: 1024,
        height: 1024,
      },
    ],
  },
  twitter: {
    handle: "@jakerunzer",
    cardType: "summary",
  },
};

export default config;
