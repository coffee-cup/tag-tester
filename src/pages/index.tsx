import * as React from "react";
import styled from "@emotion/styled";
import css from "@styled-system/css";
import Layout from "../components/Layout";
import Input from "../components/Input";

const StyledHome = styled.div(css({}));

const Home = () => (
  <Layout>
    <StyledHome>
      <h1>Open Graph Tester</h1>

      <Input value="This is large" large />
      <Input value="This is the value" />
      <Input placeholder="Enter the value here" />
    </StyledHome>
  </Layout>
);

export default Home;
