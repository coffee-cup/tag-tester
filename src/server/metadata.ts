import * as htmlparser2 from "htmlparser2";
import { MetaTag } from "../types";
import fetch from "node-fetch";
import { prefixUrl } from "../utils";

export const parseTags = (html: string): MetaTag[] => {
  const tags: MetaTag[] = [];

  let readingTitle = false;
  let inHead = false;

  const htmlParser = new htmlparser2.Parser({
    onopentag(name, attribs) {
      if (name === "meta") {
        const { name, property, content, value } = attribs;

        if (name || property) {
          tags.push({ name, property, content, value });
        }
      } else if (name === "title" && inHead) {
        readingTitle = true;
      } else if (name === "head") {
        inHead = true;
      }
    },
    ontext(text) {
      if (readingTitle) {
        tags.push({ name: "title", content: text });
      }
    },
    onclosetag(name) {
      if (name === "title") {
        readingTitle = false;
      } else if (name === "head") {
        inHead = false;
      }
    },
  });

  htmlParser.write(html);
  htmlParser.end();

  return tags;
};

export const getMetadata = async (page: string): Promise<MetaTag[]> => {
  const html = await fetch(prefixUrl(page)).then(res => res.text());

  const tags = parseTags(html);

  return tags;
};
