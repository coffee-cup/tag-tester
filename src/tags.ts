import { MetaTag, FilterType } from "./types";
import { rootUrl } from "./utils";

export const getNameProp = (tag: MetaTag): string =>
  tag.name != null ? "name" : "property";

export const getValueProp = (tag: MetaTag): string =>
  tag.content != null ? "content" : "value";

export const getName = (tag: MetaTag): string => tag.name ?? tag.property;

export const getValue = (tag: MetaTag): string => tag.content ?? tag.value;

export const htmlTags = ["title", "description", "image"];

export const twitterPrefix = "twitter:";
export const ogPrefix = "og:";

export const isTwitterTag = (tag: MetaTag): boolean => {
  const name = getName(tag);
  return name != null && name.startsWith(twitterPrefix);
};

export const isOGTag = (tag: MetaTag): boolean => {
  const name = getName(tag);
  return name != null && name.startsWith(ogPrefix);
};

export const isHTMLTag = (
  tag: MetaTag,
  onlyRecommended: boolean = true,
): boolean => {
  const name = getName(tag);
  const isRecommended =
    tag.created || (onlyRecommended ? htmlTags.includes(name) : true);
  return name != null && !isOGTag(tag) && !isTwitterTag(tag) && isRecommended;
};

const imageRegex = /^.*:?image$/;
export const isImageTag = (tag: MetaTag): boolean => {
  const name = getName(tag);
  return imageRegex.test(name);
};

export const makeTag = (
  name: string,
  content: string | undefined,
  created?: boolean,
): MetaTag => {
  const newTag: MetaTag = {
    [name.startsWith(ogPrefix) ? "property" : "name"]: name,
    [name.startsWith(twitterPrefix) ? "value" : "content"]: content,
  };

  if (created) {
    newTag.created = created;
  }

  return newTag;
};

export const createCustomUrl = (
  url: string,
  customTags: { [key: string]: string | undefined },
): string => {
  const query = {
    url,
    ...customTags,
  };

  const queryString = Object.keys(query)
    .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(query[k])}`)
    .join("&");

  return `${rootUrl}/api/page?${queryString}`;
};

export const getFilteredTags = (
  tags: MetaTag[],
  filters: FilterType[],
  onlyRecommended: boolean,
): MetaTag[] =>
  tags.filter(t => {
    if (isHTMLTag(t, onlyRecommended)) {
      return filters.includes(FilterType.Html);
    } else if (isOGTag(t)) {
      return filters.includes(FilterType.OpenGraph);
    } else if (isTwitterTag(t)) {
      return filters.includes(FilterType.Twitter);
    }

    return false;
  });

export const getBase = (s: string): string =>
  s.replace(/^og:/, "").replace(/^twitter:/, "");

export const editTagFromTags = (
  name: string,
  value: string,
  tags: MetaTag[],
  syncSimilar: boolean = false,
): {
  newTags: MetaTag[];
  edited: { [key: string]: string };
} => {
  const edited: { [key: string]: string } = {};
  const newTags = tags.map(t => {
    if (
      t[getNameProp(t)] === name ||
      (syncSimilar && getBase(name) === getBase(t[getNameProp(t)]))
    ) {
      edited[getName(t)] = value;

      return {
        ...t,
        [getValueProp(t)]: value,
      };
    }

    return t;
  });

  return {
    newTags,
    edited,
  };
};

export const deleteTagFromTags = (
  tag: MetaTag,
  tags: MetaTag[],
): {
  newTags: MetaTag[];
  edited: { [key: string]: string | undefined };
} => {
  const newTags = tags.filter(t => t !== tag);
  const edited = {
    [getName(tag)]: undefined,
  };

  return { newTags, edited };
};

export const createNewTag = (
  name: string,
  content: string,
  tags: MetaTag[],
): {
  newTags: MetaTag[];
  edited: { [key: string]: string };
} => {
  const newTag = makeTag(name, content, true);
  const edited = {
    [name]: content,
  };

  return {
    newTags: [...tags, newTag],
    edited,
  };
};
