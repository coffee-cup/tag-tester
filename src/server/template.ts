import { MetaTag } from "../types";
import { getNameProp, getValueProp } from "../tags";

export const createPage = (head: string, body: string): string =>
  `
<html lang="en">

  <head>
    <meta charset="utf-8" />
${head}

  <style>
body {
  font-family: system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", sans-serif;
  font-weight: normal;
  color: #333;
  margin: 0;
  padding: 2rem;
}

main {
  display: flex;
  alignItems: center;
  justifyContent: center;
}

.content {
  margin: 0 auto;
  maxWidth: 62rem;
}

.url {
  color: #045aff;
}
  </style>

  </head>

  <body>
    <main>
      <div class="content">
${body}
      </div>
    </main>
  </body>

</html>
`;

const generateSingleTag = (tag: MetaTag): string => {
  if (tag.name === "title") {
    return `    <title>${tag.content}</title>`;
  }

  const nameProp = getNameProp(tag);
  const valueProp = getValueProp(tag);

  return `    <meta ${nameProp}="${tag.name ??
    tag.property}" ${valueProp}="${tag.content ?? tag.value}" /> `;
};

export const generateTagPage = (
  url: string,
  tags: MetaTag[],
  customTags: string[],
): string => {
  const meta: string = tags
    .filter(t => !(t.content == null && t.value == null))
    .map(t => generateSingleTag(t))
    .join("\n");

  const tagItems = customTags.map(t => `<li>${t}</li>`).join("\n");
  const tagMessage =
    customTags.length === 0
      ? "The original tags are being used as none have been modified"
      : `
Original metatags from <a href="${url}">${url}</a> have been combined with these modified
tags

${tagItems}
`;

  const debuggerLinks = `
Enter this url on the following metatag testers

<ul>
  <li><a href="https://cards-dev.twitter.com/validator">Twitter card validator</a></li>
  <li><a href="https://developers.facebook.com/tools/debug/">Facebook Debugger</a></li>
  <li><a href="https://www.linkedin.com/post-inspector/inspect/">LinkedIn insepctor</a></li>
</ul>
`;

  const body = `
<h1>Tag Tester page for <code class="url">${url}</code></h1>

<p><a href="https://tagtester.dev?url=${encodeURIComponent(
    url,
  )}">View on tagtester.dev</a>
</p>

<p>
${tagMessage}
</p>

<p>
${debuggerLinks}
</p>
`;

  const html = createPage(meta, body);
  return html;
};
