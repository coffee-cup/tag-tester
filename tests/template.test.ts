import { generateTagPage } from "../src/server/template";
import { makeTag } from "../src/tags";

describe("html template", () => {
  it("can generate page from tags", () => {
    const tags = [makeTag("title", "a title"), makeTag("og:image", "a image")];
    const page = generateTagPage("https://tagtester.dev", tags, ["title"]);

    expect(page).toMatchSnapshot();
  });

  it("can generate page from tags with no custom", () => {
    const tags = [makeTag("title", "a title"), makeTag("og:image", "a image")];
    const page = generateTagPage("https://tagtester.dev", tags, []);

    expect(page).toMatchSnapshot();
  });

  it("can generated page with deleted tags", () => {
    const tags = [makeTag("title", undefined), makeTag("og:image", "a image")];
    const page = generateTagPage("https://tagtester.dev", tags, ["title"]);

    expect(page).toMatchSnapshot();
  });
});
