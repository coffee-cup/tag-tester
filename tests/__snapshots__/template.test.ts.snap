// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html template can generate page from tags 1`] = `
"
<html lang=\\"en\\">

  <head>
    <meta charset=\\"utf-8\\" />
    <title>a title</title>
    <meta property=\\"og:image\\" content=\\"a image\\" /> 

  <style>
body {
  font-family: system-ui, -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", sans-serif;
  font-weight: normal;
  margin: 0;
  padding: 2rem;
}

main {
  display: flex;
  alignItems: center;
  justifyContent: center;
}

.content {
  margin: 0 auto;
  maxWidth: 62rem;
}

.url {
  color: #045aff;
}
  </style>

  </head>

  <body>
    <main>
      <div class=\\"content\\">

<h1>Tag Tester page for <code class=\\"url\\">https://tagtester.dev</code></h1>

<p><a href=\\"https://tagtester.dev?url=https%3A%2F%2Ftagtester.dev\\">View on tagtester.dev</a>
</p>

<p>

Original metatags from <a href=\\"https://tagtester.dev\\">https://tagtester.dev</a> have been combined with these modified
tags

<li>title</li>

</p>

<p>

Enter this url on the following metatag testers

<ul>
  <li><a href=\\"https://cards-dev.twitter.com/validator\\">Twitter card validator</a></li>
  <li><a href=\\"https://developers.facebook.com/tools/debug/\\">Facebook Debugger</a></li>
  <li><a href=\\"https://www.linkedin.com/post-inspector/inspect/\\">LinkedIn insepctor</a></li>
</ul>

</p>

      </div>
    </main>
  </body>

</html>
"
`;

exports[`html template can generate page from tags with no custom 1`] = `
"
<html lang=\\"en\\">

  <head>
    <meta charset=\\"utf-8\\" />
    <title>a title</title>
    <meta property=\\"og:image\\" content=\\"a image\\" /> 

  <style>
body {
  font-family: system-ui, -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, \\"Helvetica Neue\\", sans-serif;
  font-weight: normal;
  margin: 0;
  padding: 2rem;
}

main {
  display: flex;
  alignItems: center;
  justifyContent: center;
}

.content {
  margin: 0 auto;
  maxWidth: 62rem;
}

.url {
  color: #045aff;
}
  </style>

  </head>

  <body>
    <main>
      <div class=\\"content\\">

<h1>Tag Tester page for <code class=\\"url\\">https://tagtester.dev</code></h1>

<p><a href=\\"https://tagtester.dev?url=https%3A%2F%2Ftagtester.dev\\">View on tagtester.dev</a>
</p>

<p>
The original tags are being used as none have been modified
</p>

<p>

Enter this url on the following metatag testers

<ul>
  <li><a href=\\"https://cards-dev.twitter.com/validator\\">Twitter card validator</a></li>
  <li><a href=\\"https://developers.facebook.com/tools/debug/\\">Facebook Debugger</a></li>
  <li><a href=\\"https://www.linkedin.com/post-inspector/inspect/\\">LinkedIn insepctor</a></li>
</ul>

</p>

      </div>
    </main>
  </body>

</html>
"
`;
